webapp/components/survey/NodeDefDetails/store/useNodeDefDetails.js

```diff
@@ -20,20 +20,33 @@ export const useNodeDefDetails = () => {
   const surveyCycleKey = useSurveyCycleKey()
 
   const [state, setState] = useState({})
+  const stateRef = useRef(state)
 
   const Actions = useActions({ setState })
 
   const editingFromDesigner = useIsDesignerNodeDefRoute()
 
   useEffect(() => {
     // Editing a nodeDef
-    if (nodeDefUuid) {
+    if (nodeDefUuid && !State.getNodeDef(state)) {
       const nodeDefSurvey = Survey.getNodeDefByUuid(nodeDefUuid)(survey)
       const validation = Survey.getNodeDefValidation(nodeDefSurvey)(survey)
       setState(State.create({ nodeDef: nodeDefSurvey, validation }))
     }
+
+    return () => {
+      // get local state from stateRef to get it up to date in this callback
+      if (State.isDirty(stateRef.current)) {
+        Actions.cancelEdits({ state: stateRef.current, showConfirm: false, goBack: false })
+      }
+    }
   }, [])
 
+  useEffect(() => {
+    // update state ref on local state update
+    stateRef.current = state
+  }, [state])
```

Comment:
why keeping a ref to the local state and updating it every time the state changes?

Reply:
see comment to line 36